test_runner: support object property mocking

idango10 opened the pull request:
<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

Fixes: https://github.com/nodejs/node/issues/58322

---

This PR adds object property mocking to the test runner.

The returned mock object provides the following methods:
- `callGetterCount()`: Retrieves the number of times the property was read.
- `callSetterCount()`: Retrieves the number of times the property was written.
- `callCount()`: Retrieves the total number of times the property was accessed (getter or setter).
- `mockValue(value)`: Sets the current mock value.
- `resetCallGetters()`: Resets only the getter call count.
- `resetCallSetters()`: Resets only the setter call count.
- `resetCalls()`: Resets both getter and setter call counts.
- `restore()`: Restores the original property value and descriptor.

**Example usage:**
```js
test('mocks a property value', (t) => {
  const obj = { foo: 42 };
  const valueMock = t.mock.value(obj, 'foo', 100);

  assert.strictEqual(obj.foo, 100);
  assert.strictEqual(valueMock.callGetterCount(), 1);

  obj.foo = 200;
  assert.strictEqual(obj.foo, 200);
  assert.strictEqual(valueMock.callSetterCount(), 1);

  valueMock.restore();
  assert.strictEqual(obj.foo, 42);
});
```

cjihrig commented on doc/api/test.md:
````diff
@@ -2240,6 +2240,47 @@ test('mocks a builtin module in both module systems', async (t) => {
 });
 ```
 
+### `mock.value(object, valueName, value)`
````
Is there any reason to make `value` required? In some of the other mocking functions, we make this optional so that we can spy on the existing value instead of needing to provide a new one.

cjihrig commented on doc/api/test.md:
```diff
@@ -2240,6 +2240,47 @@ test('mocks a builtin module in both module systems', async (t) => {
+-->
+
+* `object` {Object} The object whose value is being mocked.
+* `valueName` {string|symbol} The identifier of the value on object to mock.
```
```suggestion
* `propertyName` {string|symbol} The identifier of the property on `object` to mock.
```

cjihrig commented on doc/api/test.md:
```diff
@@ -2240,6 +2240,47 @@ test('mocks a builtin module in both module systems', async (t) => {
+
+* `object` {Object} The object whose value is being mocked.
+* `valueName` {string|symbol} The identifier of the value on object to mock.
+* `value` {any} A value used as the mock value for `object[valueName]`.
```
The return value is missing here. See the docs for `mock.method()` for an example.

cjihrig commented on doc/api/test.md:
````diff
@@ -2240,6 +2240,47 @@ test('mocks a builtin module in both module systems', async (t) => {
+});
+```
+
+The returned mock object provides the following methods:
````
`MockValueContext` needs its own section in the docs instead of being mentioned here. See `MockFunctionContext` for an example.

cjihrig commented on doc/api/test.md:
```diff
@@ -2240,6 +2240,47 @@ test('mocks a builtin module in both module systems', async (t) => {
+
+The returned mock object provides the following methods:
+
+* `callGetterCount()`: Retrieves the number of times the property was read.
+* `callSetterCount()`: Retrieves the number of times the property was written.
+* `callCount()`: Retrieves the total number of times the property was accessed (getter or setter).
+* `mockValue(value)`: Sets the current mock value.
+* `resetCallGetters()`: Resets only the getter call count.
+* `resetCallSetters()`: Resets only the setter call count.
+* `resetCalls()`: Resets both getter and setter call counts.
+* `restore()`: Restores the original property value and descriptor.
```
I think this API should be reworked to be more like the one of `MockFunctionContext` so that there is a more consistent user experience. I'm also not sure we need so many different small APIs here.

I'd definitely like to see something resembling `MockFunctionContext.calls` here. We might name it something different since these aren't methods (maybe `accesses`?). That way, we get more than just the count of accesses. We could see the actual call stacks for each access, as well as if it was a read or write, etc.

cjihrig commented on test/parallel/test-runner-mocking.js:
```diff
@@ -1054,3 +1054,68 @@ test('setter() fails if getter options is true', (t) => {
   }, /The property 'options\.setter' cannot be used with 'options\.getter'/);
 });
+
+test('spies on a property value', (t) => {
```
I think we need more tests than this. For example, we need to ensure that mocks created with this new API on the `TestContext` are automatically restored at the end of the test.

idango10 commented:
> Thanks for working on this. I think this is a good start, but I'd like to see it more unified with the rest of the mocking framework (particularly `mock.method()` since the use case is so similar)

Thanks for the thorough review! 😃 
I think I've addressed all your comments and pushed the changes. Let me know if there's anything else you'd like me to tweak.

benjamingr commented:
Hey, welcome @idango10 nice seeing you around here, I trust Colin and agree with his feedback.

cjihrig reviewed (commented):
This is looking pretty good to me. Two notes:

1. It doesn't appear that a `mockImplementationOnce()` equivalent is implemented.
2. Can we add a test for explicitly mocking a property as `undefined`. This will definitely come up eventually, and I don't think the current implementation can support it. It might take a little bit of work to distinguish between the "explicit `undefined`" and "use the previously existing value" use cases.

idango10 commented:
> This is looking pretty good to me. Two notes:
> 
> 1. It doesn't appear that a `mockImplementationOnce()` equivalent is implemented.
> 2. Can we add a test for explicitly mocking a property as `undefined`. This will definitely come up eventually, and I don't think the current implementation can support it. It might take a little bit of work to distinguish between the "explicit `undefined`" and "use the previously existing value" use cases.

1. I'm wondering about how we should implement this. It's pretty problematic since `accesses` includes both get and set actions. For gets we have no problem, but set is a different case. If a set happens at the same access index, it will consume the "once" value and change the mocked property to the value we only intended to use once.
2. Great catch, it was indeed a bug. Fixed and added a test for it.

ljharb commented:
@peteduel because enumerability and writability are all things that should be tested?
